### ISSUE TITLE
Ability to reset the ScriptHost for allowing multiple runs on the same process

### ISSUE THREAD

**Issue by augustoproiete** [devlead/Cake.Bridge#1]
I'm experimenting with Cake.Bridge on a custom host that has the ability to run Cake tasks multiple times on the same process (e.g. Windows client app).

The first run of the Cake tasks works as expected, but any attempt to run the tasks again fail with the error `Another task with the name 'xxx' has already been added.` because the `ScriptHost` is created only once in the static context - that is still the same on the second run of the tasks.

### Repro steps

```csharp
using Cake.Common;
using Cake.Common.Diagnostics;
using Cake.Core;
using static CakeBridge;

namespace ExampleCakeBridge
{
    class Program
    {
        static void Main(string[] args)
        {
            RunCake(); // Works as expected
            RunCake(); // CakeException: Another task with the name 'Test' has already been added
        }

        private static void RunCake()
        {
            var target = Context.Argument("target", "Test");
            var configuration = Context.Argument("configuration", "Release");

            Task("Test")
                .Does(() =>
                {
                    Context.Information($"Hello from App Host ({configuration})");
                });

            RunTarget(target);
        }
    }
}
```

Re-creating the `ScriptHost` via `CakeBridge.GetScriptHost` would resolve the issue but the `ScriptHost` property is read-only, so that's not an option at the moment.

I suggest adding a `ResetScriptHost` method to explicitly replace the `ScriptHost` with a new one via `CakeBridge.GetScriptHost`.

I'm happy to send a PR.

**Comment by devlead** [comment]
I'm not 100% with on the use case.

ScriptHost will never be thread-safe if it can be exchanged at any time, this won't work in a regular Cake either.

If it's the same tasks that are going to be executed, then you could always Setup once and RunTarget multiple times
```csharp
var target = Context.Argument("target", "Test");
var configuration = Context.Argument("configuration", "Release");

Task("Test")
    .Does(() =>
    {
        Context.Information($"Hello from App Host ({configuration})");
    });

RunTarget(target);
RunTarget(target);
```

To make clear one could perhaps do something like
```csharp
public async static Task<int> ExecuteScopedAsync(Func<Task<int>> action)
public static int ExecuteScoped(Func<int> action)
```

Then code would look something like
```csharp
using Cake.Common;
using Cake.Common.Diagnostics;
using Cake.Core;
using static CakeBridge;

namespace ExampleCakeBridge
{
    class Program
    {
        static void Main(string[] args)
        {
            ExecuteScoped(RunCake);
            ExecuteScoped(RunCake);
        }

        private static void RunCake()
        {
            var target = Context.Argument("target", "Test");
            var configuration = Context.Argument("configuration", "Release");

            Task("Test")
                .Does(() =>
                {
                    Context.Information($"Hello from App Host ({configuration})");
                });

            RunTarget(target);
        }
    }
}
```

Still wouldn't be thread safe unless guards were put in place, naive implementation of above something like
```csharp
  public async static Task ExecuteScopedAsync(Func<Task> action)
    {
        try
        {
             await action();
        }
        finally
        {
            ScriptHost = GetScriptHost();
        }
    }

    public static void ExecuteScoped(Action action)
     {
        lock(ScriptHost)
        {
            try
            {
                action();
            }
            finally
            {
                ScriptHost = GetScriptHost();
            }
        }
    }
```

**Comment by augustoproiete** [comment]
Use-case: I have a Windows client app that references `Cake.Bridge` and the user can drag & drop one or more `.cs` files to it that would have some code like the example above:

``` 
var target = Context.Argument("target", "Test");
var configuration = Context.Argument("configuration", "Release");

Task("Test")
    .Does(() =>
    {
        Context.Information($"Hello from App Host ({configuration})");
    });

RunTarget(target);
```

I have no control over the code that is brought to the app, and my app is really just a runner. When the user clicks on a button to run the code, I'm doing something similar to your `ExecuteScoped` / `ExecuteScopedAsync` example above, except that I cannot set a new value for `ScriptHost` because it's read-only.

Having a way of creating a new `ScriptHost` would remove the need for creating a new process every time the user tries to run some code.

**Comment by devlead** [comment]
My thinking was if `ScriptHost` got a private setter and it `ExecuteScoped` / `ExecuteScopedAsync` was added to `CakeBridge`.

So it's only through opt-in using those methods this behavior would be introduced.

**Comment by augustoproiete** [comment]
Yes, I believe that would work.

**Comment by augustoproiete** [comment]
That said, (and as I think more about it) the ideal scenario would be to have the ability of having multiple `ScriptHost`s on the same process (e.g. `AsyncLocal`) which would allow running multiple Cake scripts on the same process at the same time, in separate threads.

A thread-safe `ExecuteScoped` / `ExecuteScopeAsync` would be a good step forward though - and address what I originally asked above.

**Comment by devlead** [comment]
You could always call `public static IScriptHost GetScriptHost()` and work using that instance.

I.e.
```csharp
using Cake.Common;
using Cake.Common.Diagnostics;
using Cake.Core;

namespace ExampleCakeBridge
{
    class Program
    {
        static void Main(string[] args)
        {
            RunCake(); // Works as expected
            RunCake(); // Works as expected too
        }

        private static void RunCake()
        {
            var host = CakeBridge.GetScriptHost();

            var target = host.Context.Argument("target", "Test");
            var configuration = host.Context.Argument("configuration", "Release");

            host.Task("Test")
                .Does(context =>
                {
                    context.Information($"Hello from App Host ({configuration})");
                });

            host.RunTarget(target);
        }
    }
}
```

**Comment by augustoproiete** [comment]
Sure, but having to specify the host on every call takes away a huge part of the attractiveness of Cake.Bridge (to me anyway)

With the regular usage of Cake.Bridge, users can copy/paste the exact same code on a `.cake` file, run `dotnet cake` and It Just Works™... The code doesn't know about `Cake.Bridge` or `ScriptHost`, etc. and I think this is super cool.

If devs need to specify the `host` on every call, the code is no longer a "Cake Script" code, but instead it's a "Cake Bridge" code, if that makes sense.

**Comment by devlead** [comment]
Ok, get it, would you be willing to send a PR?

**Comment by augustoproiete** [comment]
Yes, I'll take a stab at it. Cheers.
